incubator/hnc/pkg/validators/config.go

```diff
@@ -0,0 +1,182 @@
+// +kubebuilder:webhook:path=/validate-hnc-x-k8s-io-v1alpha1-hncconfigurations,mutating=false,failurePolicy=fail,groups="hnc.x-k8s.io",resources=hncconfigurations,verbs=create;update;delete,versions=v1alpha1,name=hncconfigurations.hnc.x-k8s.io
+
+type Config struct {
+	Log     logr.Logger
+	client  validationClient
+	decoder *admission.Decoder
+}
+
+// validationClient checks if a resource exists. The check should typically be performed against the apiserver,
+// but need to be stubbed out during unit testing.
+type validationClient interface {
+	// ResourceExists takes a GVK, and returns true if the corresponding resource of the GVK exists in the apiserver.
+	ResourceExists(ctx context.Context, gvk schema.GroupVersionKind) (bool, error)
+}
+
+type set map[schema.GroupVersionKind]bool
+
+func (c *Config) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if isHNCServiceAccount(&req.AdmissionRequest.UserInfo) {
+		return allow("HNC SA")
+	}
+
+	if req.Operation == v1beta1.Delete {
+		if req.Name == "config" {
+			return deny(metav1.StatusReasonForbidden, "Deleting the 'config' object is forbidden")
+		} else {
+			// We allow deleting other objects. If the validation controller has always been running, we should not
+			// enter this case because objects of other names cannot be created. If users somehow bypass the
+			// validation controller and create objects of other names, we will allow them to delete the objects.
+			return allow("")
+		}
+	}
+
+	inst := &api.HNCConfiguration{}
+	if err := c.decoder.Decode(req, inst); err != nil {
+		c.Log.Error(err, "Couldn't decode request")
+		return deny(metav1.StatusReasonBadRequest, err.Error())
+	}
+
+	resp := c.handle(ctx, inst)
+	c.Log.Info("Handled", "allowed", resp.Allowed, "code", resp.Result.Code, "reason", resp.Result.Reason, "message", resp.Result.Message)
+	return resp
+}
+
+// handle implements the validation logic of this validator for Create and Update operations,
+// allowing it to be more easily unit tested (ie without constructing a full admission.Request).
+func (c *Config) handle(ctx context.Context, inst *api.HNCConfiguration) admission.Response {
+	if inst.GetName() != "config" {
+		return deny(metav1.StatusReasonForbidden, fmt.Sprintf("Wrong singleton name: %s; the name should be 'config'", inst.GetName()))
+	}
+
+	roleExist := false
+	roleBindingExist := false
+	ts := set{}
+	for _, t := range inst.Spec.Types {
+		if rp := c.isTypeConfigured(t, ts); !rp.Allowed {
+			return rp
+		}
+
+		if c.isRBAC(t, "Role") {
+			if rp := c.validateRBAC(t.Mode, "Role"); !rp.Allowed {
+				return rp
+			}
+			roleExist = true
+		} else if c.isRBAC(t, "RoleBinding") {
+			if rp := c.validateRBAC(t.Mode, "RoleBinding"); !rp.Allowed {
+				return rp
+			}
+			roleBindingExist = true
+		} else {
+			if rp := c.validateType(ctx, t, ts); !rp.Allowed {
+				return rp
+			}
+		}
+	}
+	if !roleExist {
+		return deny(metav1.StatusReasonForbidden, "Configuration for Role is missing")
```

Comment:
StatusReasonInvalid?

Reply:
Thanks for the suggestion. Changed to `StatusReasonInvalid`. Same below and the rest of the code.

Looks like for the `StatusReasonInvalid`, the webhook will only show `The request is invalid` text message when applying an invalid request. The detailed message can be found in the log. I suspect this is intended but it is less convenient than `StatusReasonForbidden`, which will show the error message after apply a request, e.g.,

> Error from server (Forbidden): admission webhook "hncconfigurations.hnc.x-k8s.io" denied the request: Deleting the 'config' object is forbidden

Reply:
I thought there was some way to send more custom text back to the user? The log isn't really good enough, IMO.

Reply:
From the documentation of [StatusReasonInvalid](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#StatusReason) we can add "cases" to the `Details` field of the [Status](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#Status). I also tried that but it still did not show the custom text. I am wondering if there is something specific about `StatusReasonInvalid` that the webhook ignore all the custom text. However, I haven't found any related logic under [controller-runtime/pkg/webhook](https://github.com/kubernetes-sigs/controller-runtime/tree/master/pkg/webhook).

We can ask the cotroller-runtime team about that to see if what might be cause of that. In the meantime, I am thinking about switching the reason back to `StatusReasonForbidden` since I feel it is the second appropriate one and it shows the custom text. WDYT?

Reply:
As we discussed offline, I changed it back to `StatusReasonForbidden` in #525. I will ask the controller runtime team to see if this behavior is expected. Thanks for the suggestion!